File: Day 1/solution.java
import java.io.*;
import java.util.*;

class DriverMain {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while(t-- > 0) {
            String txt = br.readLine().trim();
            String pat = br.readLine().trim();

            int ans = new Solution().search(pat, txt);
            System.out.println(ans);
        }
    }
}

class Solution {

    HashMap<Character, Integer> m = new HashMap<Character, Integer>();

    // Variable to check whether the count of unique letters in the map created is zero
    int count = 0;
    
    // Function to first create a map containing all the unique letters in the pattern and their frequencies
    void createPatList(String str) {
        for(char ch : str.toCharArray()) {
            if(this.m.get(ch) == null) {
                this.m.put(ch,1);
            }
            else {
                this.m.put(ch,(this.m.get(ch)+1));
            }
            if (this.m.get(ch) == 1) {
                this.count++;
            }
        }
    }
    
    // Function to deduct from the frequency of the unique letters created in the previous function
    void removeFromList(char ch) {
        if(this.m.get(ch) != null) {
            this.m.put(ch,(this.m.get(ch)-1));
            if(this.m.get(ch) == 0) {
                this.count--;
            }
        }
    }
    
    // Function to add to the frequency of unique letters
    void addToList(char ch) {
        if(this.m.get(ch) != null) {
            this.m.put(ch,(this.m.get(ch)+1));
            if (this.m.get(ch) == 1) {
                this.count++;
            }
        }
    }
    
    int search(String pat, String txt) {
        
        this.createPatList(pat);
        int ans = 0;
        int i = 0, j = 0;
        int k = pat.length();
        int n = txt.length();
        while(j<n){
            this.removeFromList(txt.charAt(j));
            if(j-i+1 < k){
                j++;
            }
            else{
                if(j-i+1 == k){
                    if (this.count == 0){
                        ans++;
                    }
                    this.addToList(txt.charAt(i));
                    i++;
                    j++;
                }
            }
            
        }
        return ans;
    }
}

File: Day 1/solution.js
/**
 * @param {string} pat
 * @param {string} txt
 * @return {number}
*/

class Solution {
    
    search(pat,txt){
       
       function createPatList(str) {
           for(var ch of str) {
               // Create the list
               if(m.get(ch) === undefined) {
                m.set(ch,1);
            }
            else {
                m.set(ch,(m.get(ch)+1));
            }
            if (m.get(ch) === 1) {
                count++;
            }
           }
       }
       
       function removeFromList(ch) {
        if(m.get(ch) !== undefined) {
            m.set(ch,(m.get(ch)-1));
            if(m.get(ch) == 0) {
                count--;
            }
        }
       }
       
       function addToList(ch) {
        if(m.get(ch) !== undefined) {
            m.set(ch,(m.get(ch)+1));
            if (m.get(ch) == 1) {
                count++;
                }
            }
        }
    
        let m = new Map();
        let count = 0;
        createPatList(pat);
        let ans = 0;
        let i = 0, j = 0;
        let k = pat.length;
        let n = txt.length;
        while(j<n){
            removeFromList(txt.charAt(j));
            if(j-i+1 < k){
                j++;
            }
            else{
                if(j-i+1 === k){
                    if (count === 0){
                        ans++;
                    }
                    addToList(txt.charAt(i));
                    i++;
                    j++;
                }
            }
            
        }
        return ans;
    }
}